### PR TITLE
issue: 964360 Improve Page Fault event on mlx5_send call

### PR DESCRIPTION
**Pull request by igor-ivanov**
These changes designed to eliminate page fault event during
qp_mgr::mlx5_send() call.

See related at master https://github.com/Mellanox/libvma/pull/292

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/924/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/776/ for details (Mellanox internal link).

File: src/vma/dev/qp_mgr.cpp [hunk 1]
```diff
@@ -536,15 +554,21 @@ void qp_mgr::mlx5_send(vma_ibv_send_wr *p_send_wqe)
 	 * implementations may use move-string-buffer assembler instructions,
 	 * which do not guarantee order of copying.
 	 */
-	COPY_64B_NT(dst, m_sq_hot_wqe);
+	mlx5_bf_copy((unsigned long long *)((uintptr_t)m_sq_bf_reg + m_sq_bf_offset),
```
Comment (OphirMunk) [inline src/vma/dev/qp_mgr.cpp]: @igor-ivanov 
1. Why don't you use *volatile* as in the original code? 
2. What is the reason for changing the casting from uintptr_t to unsigned long long? 

Can you restore both volatile and casting?

File: src/vma/dev/qp_mgr.cpp [hunk 2]
```diff
@@ -536,15 +554,21 @@ void qp_mgr::mlx5_send(vma_ibv_send_wr *p_send_wqe)
...
+		(unsigned long long *)m_sq_hot_wqe);
+
 	m_sq_bf_offset ^= m_sq_bf_buf_size;
 
+	m_sq_wqe_idx_to_wrid[m_sq_hot_wqe_index] = (uintptr_t)p_send_wqe->wr_id;
+
+	++m_sq_wqe_counter;
```
Comment (OphirMunk) [inline src/vma/dev/qp_mgr.cpp]: In the original code m_sq_wqe_counter was incremented earlier. Why do you delay its increment to this position?

File: src/vma/dev/qp_mgr.cpp [hunk 3]
```diff
@@ -500,34 +512,40 @@ int qp_mgr::post_recv(mem_buf_desc_t* p_mem_buf_desc)
...
-	++m_sq_wqe_counter;
-	volatile uintptr_t *dst = (volatile uintptr_t *)
-			((uintptr_t)m_sq_bf_reg + m_sq_bf_offset);
-
+	/*
+	 * Make sure that descriptors are written before
+	 * updating doorbell record and ringing the doorbell
+	 */
 	wmb();
 	*m_sq_db = htonl(m_sq_wqe_counter);
```
Comment (OphirMunk) [inline src/vma/dev/qp_mgr.cpp]: In the original code m_sq_wqe_counter is already incremented while in your PR it is not. 
Please explain if 
```
*m_sq_db = htonl(m_sq_wqe_counter);
```
has the same value before/after your PR. Please justify this logic

Comment (igor-ivanov) [inline src/vma/dev/qp_mgr.cpp]: I think it is my mistake. I will change.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/942/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/783/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/951/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/965/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1009/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/803/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1032/ for details (Mellanox internal link).

**Comment by OphirMunk** [comment]
Merged #293

**Pull request closed without merging**